Closes #2134 add general option signif_digits

millerg23 opened the pull request:
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

millerg23 commented:
@bms63 @zdz2101 would appreciate some early feedback on this. I have created a new general option to hold significant digits, do you think we need to document it anywhere else? We mentioned it in Lab Grading Vignette but not that its a general option. I couldn't see any particular documentation around general options, except in the man files of the functions.

StefanThoma commented:
I have two suggestions: 

- We could print messages or warnings when `signif` is used when the user has not actively specified digits themselves. This way we can ensure that users are aware of our use of `signif`.

- For some functions we expect users to input comparisons themselves (I think). We should mention in the documentation for these arguments that we encourage the use of `signif`. Alternatively, we could scan their inputs and return a message or a warning when no such measures were taken (that is probably a bit more tricky, and a bit more hand-holding, but I think it would be pretty useful).

What do you all think?

zdz2101 reviewed (commented):
@millerg23 Looks good to me thus far, I think the documentation inside `set_admiral_options()` is sufficient just because I think this will be seldom used. We already have a snippet on implementation [here](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html?q=admiral_opt#admiral-options) too for developers. I think we just need to add a simple test in our test-suite to demonstrate the proper usage of it, maybe even the same example that spun this all up to begin with?

bms63 commented:
> I have two suggestions:
> 
> * We could print messages or warnings when `signif` is used when the user has not actively specified digits themselves. This way we can ensure that users are aware of our use of `signif`.
> * For some functions we expect users to input comparisons themselves (I think). We should mention in the documentation for these arguments that we encourage the use of `signif`. Alternatively, we could scan their inputs and return a message or a warning when no such measures were taken (that is probably a bit more tricky, and a bit more hand-holding, but I think it would be pretty useful).
> 
> What do you all think?


I'm okay with the first option.  @millerg23 could you implement with this PR?

The second one sounds very tough to implement to me.  Also a little invasive to the user where now admiral is watching over them :)

bms63 commented on R/admiral_options.R:
What do you think about adding an additional example here for lab stuff?

![image](https://github.com/pharmaverse/admiral/assets/10111024/950f4ed5-233c-4e48-ae8b-a2b680797ecc)

millerg23 replied:
Done.

bms63 reviewed (commented):
Does anyone know what the See Also is used here? @pharmaverse/admiral 
![image](https://github.com/pharmaverse/admiral/assets/10111024/7c94981f-f6e3-41ce-8835-157ecff451fb)

bms63 commented:
> Does anyone know what the See Also is used here? @pharmaverse/admiral ![image](https://user-images.githubusercontent.com/10111024/280103798-7c94981f-f6e3-41ce-8835-157ecff451fb.png)

ah nvm i see that this is being used in those functions.

bms63 commented:
Okay ... got another one!!  @pharmaverse/admiral  Is this statement true?

![image](https://github.com/pharmaverse/admiral/assets/10111024/6be09921-8659-4266-a4da-5fa8ff3b9c74)

bms63 commented on R/derive_var_anrind.R:
```diff
@@ -5,6 +5,10 @@
 #'
+#' @param signif_dig Number of significant digits to use when comparing values.
+#'
+#'   Significant digits used to avoid floating point discrepancies when comparing numeric values.
```
Ohhhh..maybe a quick link to the blog post on this issue?

millerg23 replied:
Done, added same link to `derive_var_atoxgr_dir()` and also Lab grading vignette.

bms63 commented:
> @bms63 @zdz2101 would appreciate some early feedback on this. I have created a new general option to hold significant digits, do you think we need to document it anywhere else? We mentioned it in Lab Grading Vignette but not that its a general option. I couldn't see any particular documentation around general options, except in the man files of the functions.

I think if we incorporate the messaging suggested by @StefanThoma, couple with the updated documentation and mention in vignette, then we have done the best that we can.   Love to see the blog post pointed to as well!!

After that - just need some tests.  Many thanks again!!

bms63 commented:
@millerg23 just checking on progress and if you need anything from me?

millerg23 commented:
@bms63 @zdz2101 @StefanThoma I think I addressed all QC comments. Added a unit test for derive_var_anrind, to show how increasing significant digits changes the result. Added links to blog post in several places included NEWS file. Also added checks that a positive integer is being passed in for significant digits in the 2 functions that use it

zdz2101 commented on tests/testthat/test-derive_var_anrind.R:
```diff
@@ -129,3 +129,36 @@ test_that("derive_var_anrind Test 6: missing `AVAL` is handled properly", {
+## Test 8: Show floating points not handled correctly ----
+test_that("derive_var_anrind Test 8: Show floating points not handled correctly", {
+  # when SIGNIFICANT DIGITS = 17 then AVAL > ANRHI
+  # even though 1.1 * 100 should equal 110
```
This is brilliant!

bms63 commented on NEWS.md:
```diff
@@ -2,6 +2,12 @@
 
 ## New Features
 
+New global option created `signif_digits` to handle floating point issue, the value is
+set to `15`, and is used with the `base R` function `signif()` when comparing 2 numeric
+values. This is implemented in `admiral ` functions `derive_var_atoxgr_dir()` and `derive_var_anrind()`.
```
```suggestion
New global option created `signif_digits` to handle floating point issue, the value is
set to `15`, and is used with the `base R` function `signif()` when comparing 2 numeric
values. This is implemented in `admiral ` functions `derive_var_atoxgr_dir()` and `derive_var_anrind()`. (#2134)
```

bms63 commented on NEWS.md:
```diff
@@ -41,6 +47,9 @@ were enhanced such that more than one summary variable can be derived, e.g.,
 allow more control of the selection of records. It creates a temporary variable
 for the event number, which can be used in `order`. (#2140)
 
+- `signif_dig` argument added to both `derive_var_atoxgr_dir()` and `derive_var_anrind()`
+functions with default value set to general option `signif_digits`. The new argument to
+these functions handles any floating point issues.
```
```suggestion

- `signif_dig` argument added to both `derive_var_atoxgr_dir()` and `derive_var_anrind()`
functions with default value set to general option `signif_digits`. The new argument to
these functions handles any floating point issues. (#2134)
```

bms63 reviewed (approved):
Just a few minor News updates.  @kaz462 or @StefanThoma did you want to do any further review of this update?

StefanThoma commented on vignettes/lab_grading.Rmd:
```diff
@@ -87,7 +87,7 @@ When comparing numeric values, for example `AVAL > 1.1*ANRHI`, unexpected result
 both side of the equation, the number of significant digits used to compare is passed into the 
 function `derive_var_atoxgr_dir()` via the argument `signif_dig`. Please see documentation of the 
-function for more details.
+function for more details and also blog: [How admiral handles floating points](https://pharmaverse.github.io/blog/posts/2023-10-30_floating_point/floating_point.html).
```
```suggestion
function for more details and the blogpost [How admiral handles floating points](https://pharmaverse.github.io/blog/posts/2023-10-30_floating_point/floating_point.html) for more context.
```

StefanThoma replied:
Looks good to me! 👍 👍 👍 👍

millerg23 commented on vignettes/lab_grading.Rmd:
```diff
@@ -87,7 +87,7 @@ When comparing numeric values, for example `AVAL > 1.1*ANRHI`, unexpected result
 both side of the equation, the number of significant digits used to compare is passed into the 
 function `derive_var_atoxgr_dir()` via the argument `signif_dig`. Please see documentation of the 
-function for more details.
+function for more details and the blogpost [How admiral handles floating points](https://pharmaverse.github.io/blog/posts/2023-10-30_floating_point/floating_point.html) for more context.
```
```suggestion
function for more details and the blog post [How admiral handles floating points](https://pharmaverse.github.io/blog/posts/2023-10-30_floating_point/floating_point.html) for more context.
```

millerg23 commented:
@bms63 @StefanThoma thanks for feedback, I added in suggested text.

bms63 commented:
![image](https://github.com/pharmaverse/admiral/assets/10111024/4450bf04-76b8-4b84-b0c9-d99f074cc089)
